mlos_bench/mlos_bench/environments/remote/remote_env.py

```diff
@@ -175,10 +176,14 @@ def _remote_exec(self, script: Iterable[str]) -> Tuple[Status, datetime, Optiona
             Status is one of {PENDING, SUCCEEDED, FAILED, TIMED_OUT}
         """
         env_params = self._get_env_params()
+        command_name = re.sub(r"\W+", "-", self.name).lower()
```

Comment:
I am not sure that using the `Environment` name as command ID will work. The environment can have separate commands for setup, run, and teardown, at least. In addition to that, other environments can inherit from `RemoteEnv` and make additional calls to `_remote_exec()` for their internal needs (e.g., early termination).

So maybe pass the command name as an additional parameter to `_remote_exec()`?

**P.S.** We also have to think about the parallelization of the experiments, when (theoretically) several trials may simultaneously run different commands on the same VM

Reply:
Agreed. For the previous behavior, setup, run and teardown would just keep overriding the managed run command.
Added the command name as an additional parameter, and now each step adds a related suffix to the name.

Reply:
> The first point is really needed for steps like the benchbase loading phase, where loading TPCC scale factor 200+ is a struggle due to hitting the previous 90 min timeout limit.

Another option here is make these operations all two phase.  e.g., they could all have a "start" and "finish" sub-phase where "start_setup" is used to kick off the loading process in the background and "finish_setup" is used to wait for its result and report that back to MLOS.

That would allow the script writer to create async operations and have polling in the background and it would work across multiple backend services (e.g., SshService).

Reply:
> P.S. We also have to think about the parallelization of the experiments, when (theoretically) several trials may simultaneously run different commands on the same VM

I'm not certain about that part.  There should only ever be one trial executing on an Environment (VM) at a time.

We *will* want to have different Environment copies for parallel Trial execution across Workers though (#720).

Reply:
> Agreed. For the previous behavior, setup, run and teardown would just keep overriding the managed run command. Added the command name as an additional parameter, and now each step adds a related suffix to the name.

Here's my proposal:

1. Assemble a json (dict) of 

```python
command_id_info = {
  "environment_id": self.id,
  "command": command_str,
  "config": config,
  "globals": globals,
}
```

  Basically it should include anything that goes into making that execution unique.
  In particular, the environment id should include something about the name of the environment and its type, the config should include the particulars for this invocation and maybe also the trial_id, the globals may incorporate any other substitutions, the command is obviously the particular command being executed, etc.

2. Create a digest of the info: `sha256(json.dumps(command_info)).hex` or somesuch

Reply:
The question about outstanding unfinished commands is import though.

Can we check for those prior to issuing a new exec?

Reply:
> 2\. Create a digest of the info: `sha256(json.dumps(command_info)).hex` or somesuch

This sounds like a good idea, although I need to confirm this fits within the length and character limits of command names.

There is currently a big blocker for including trial-specific info like the `trial_id` and `parameters` into the command naming. Currently, there is a limit of 25 managed run commands per instance: https://learn.microsoft.com/en-us/azure/virtual-machines/run-command-overview#when-to-use-action-or-managed-commands
This would severely limit how many trials we can run, even if assuming each trial only makes one remote exec call.


> The question about outstanding unfinished commands is import though.
> 
> Can we check for those prior to issuing a new exec?

Yes this can be done, either by issuing a GET on the endpoints `runcommands` or `runcommands/{command_name}`, where we can see the execution status. I think it is a good idea to check unfinished commands before issuing a new one, but maybe better done in a separate PR as we don't support that even currently.

Reply:
> This sounds like a good idea, although I need to confirm this fits within the length and character limits of command names.

sha256 produces a 64 character hex hash.  Just truncate it to fit whatever the limit is.  git uses sha1 and often only reports the first 7 characters.  That still gives (26+10)^7 ~ 78 billion possibilities.  It doesn't need to be perfect.  Probabilistic is still very good for this case since commands will eventually be retired (completed and removed from the history on the backend) over time.

Reply:
> There is currently a big blocker for including trial-specific info like the trial_id and parameters into the command naming. Currently, there is a limit of 25 managed run commands per instance: https://learn.microsoft.com/en-us/azure/virtual-machines/run-command-overview#when-to-use-action-or-managed-commands
This would severely limit how many trials we can run, even if assuming each trial only makes one remote exec call.

I see the issue now.  Managed run scripts are a resource you create and associate with a VM and then invoke.

So, to work around the limits, the workflow would have to be something like:

1. Create managed script resource.
2. Invoke it.
3. Wait for the result.
4. Delete the managed script resource.

That adds some additional back and forth with ARM which can be slow unfortunately.

Question:

Can we we create somewhat more generic managed scripts objects that can be invoked with arguments and/or environment variables (or maybe fed via a file) in order to adjust their behavior.
*If* there are less than 25 of them, then we can create the objects once and execute as necessary.

Reply:
Actually, given the output blob URIs are also fixed, we might have to do this create/invoke/wait/delete operation every time anyways.